### PR TITLE
fix(forms-web-app): controller, validator, route + unit tests for eligibility listed building

### PR DESCRIPTION
**Pull request by a6software**
## Ticket Number

https://pins-ds.atlassian.net/browse/AS-1858

## Description of change
Fixes issue whereby given answer was not being saved or redisplayed on the front end

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `master` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

**Comment by a6software** [comment]
I think this might fail until the appeal service api change is in (?)

File: packages/forms-web-app/src/controllers/eligibility/listed-building.js [hunk 1]
```diff
@@ -1,46 +1,62 @@
...
     });
     return;
   }
 
-  if (body['is-your-appeal-about-a-listed-building'] === 'yes') {
+  try {
+    req.session.appeal = await createOrUpdateAppeal({
+      ...appeal,
+      eligibility: {
+        ...appeal.eligibility,
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/controllers/eligibility/listed-building.js:44]: Does this duplicate code entered in line 42? Not duplicating is probably more inconvenient...

File: packages/forms-web-app/src/controllers/eligibility/listed-building.js [hunk 2]
```diff
@@ -1,46 +1,62 @@
...
+        isListedBuilding: isAppealAboutAListedBuilding,
+      },
+    });
+  } catch (e) {
+    logger.error(e);
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/controllers/eligibility/listed-building.js:49]: Is this in keeping with the new logging conventions recently updated in Confluence?

File: packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js [hunk 3]
```diff
@@ -19,54 +35,100 @@ describe('controllers/eligibility/listed-building', () => {
...
+          'is-your-appeal-about-a-listed-building': 'bad value',
+          errors: { a: 'b' },
+          errorSummary: [{ text: 'There were errors here', href: '#' }],
+        },
       };
-      const mockResponse = {
-        redirect: jest.fn(),
+      await listedBuildingController.postListedBuilding(mockRequest, res);
+
+      expect(createOrUpdateAppeal).not.toHaveBeenCalled();
```
Comment (swdpcomputing) [inline packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js:56]: This is not testing re-rendering, should this be in it's own explicitly named test?

File: packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js [hunk 4]
```diff
@@ -19,54 +35,100 @@ describe('controllers/eligibility/listed-building', () => {
...
+
+      expect(res.redirect).not.toHaveBeenCalled();
```
Comment (swdpcomputing) [inline packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js:58]: As last comment

File: packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js [hunk 5]
```diff
@@ -19,54 +35,100 @@ describe('controllers/eligibility/listed-building', () => {
...
+        body: {},
       };
 
-      listedBuildingController.postListedBuilding(mockRequest, mockResponse);
+      const error = new Error('Cheers');
+      createOrUpdateAppeal.mockImplementation(() => Promise.reject(error));
+
+      await listedBuildingController.postListedBuilding(mockRequest, res);
+
+      expect(res.redirect).not.toHaveBeenCalled();
```
Comment (swdpcomputing) [inline packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js:83]: As last comment

**Review by rcmsoftwaredevelopment — APPROVED** [review]

File: packages/forms-web-app/src/controllers/eligibility/listed-building.js (hunk 1 above)
Comment (a6software) [reply packages/forms-web-app/src/controllers/eligibility/listed-building.js:44]: Line 44 is required or the existing key / values inside `eligibility` would be lost.

Perhaps better illustrated with a screenshot:

Screenshot from 2021-04-12 08-56-28

`a` is the original object

`x` is a wholesale copy

`y` is a copy that redefines its own `b`, but in doing so we lose the contents of `a.b`.

`p` demonstrates the additional spread (like line 44) which retains everything that exists in `a.b` and adds in new stuff, too.

File: packages/forms-web-app/src/controllers/eligibility/listed-building.js (hunk 2 above)
Comment (a6software) [reply packages/forms-web-app/src/controllers/eligibility/listed-building.js]: I believe I have now resolved this. Thanks for flagging!

File: packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js (hunk 3 above)
Comment (a6software) [reply packages/forms-web-app/tests/unit/controllers/eligibility/listed-building.test.js:54]: These steps (this and the other two comments that relate) are testing the function calls that happen before the re-rendering. If I split them out I would have something like this:

(assume we have five steps in the controller)

test 1:

1. controller call
2. sub func 1 is called

test 2:

1. controller call
2. sub func 1 is called
3. sub func 2 is called

test 2:

1. controller call
2. sub func 1 is called
3. sub func 2 is called
4. sub func 3 is called

test 2:

1. controller call
2. sub func 1 is called
3. sub func 2 is called
4. sub func 3 is called
5. sub func 4 is called

test 2:

1. controller call
2. sub func 1 is called
3. sub func 2 is called
4. sub func 3 is called
5. sub func 4 is called
6. sub func 5 is called

Seems like a lot of repeated tests to get to the same outcome. 

But equally, I feel like I have misunderstood?

**Review by rcmsoftwaredevelopment — APPROVED** [review]

**Comment by a6software** [comment]
I am merging this as it's blocking two tickets and I am happy this approach follows the existing approach re: the three test comments.

@swdpcomputing I will address any feedback under a new PR as needed.

**Comment by swdpcomputing** [comment]
LGTM

**Comment by a6software** [comment]
thanks!

**Pull request merged**